### PR TITLE
Make copy of instance's devices before executing changes to system

### PR DESCRIPTION
**Pull request by TomasKorbar**
- these changes will protect tuned's execution thread from errors
caused by changes of instance's devices list while execution
thread iterates over them
- fixes problem with bad recognition of disk caused
by use of python3
- Resolve: rhbz#1592743

### CODE REVIEW

File: tuned/plugins/base.py [hunk 1]
```diff
@@ -40,6 +41,7 @@ def __init__(self, monitors_repository, storage_factory, hardware_inventory, dev
 		self._options_used_by_dynamic = self._get_config_options_used_by_dynamic()
 
 		self._cmd = commands()
+		self._current_devices = None
```
[the remark this answers is not in the document]
Comment (TomasKorbar) [reply tuned/plugins/base.py]: Yes. List of devices is now passed from important methods like instance_apply_tuning to methods which are working with it.

File: tuned/plugins/base.py [hunk 2]
```diff
@@ -236,19 +238,24 @@ def _call_device_script(self, instance, script, op, devices, full_rollback = Fal
 				ret = False
 		return ret
 
+	def _refresh_current_devices(self, instance):
+		self._current_devices = copy.deepcopy(instance.devices)
```
[the remark this answers is not in the document]
Comment (TomasKorbar) [reply tuned/plugins/base.py]: Good point. A shallow copy is sufficient.

**Comment by olysonek** [comment]
To give an update here, during review of this PR I realized that in the current codebase there are a bunch of other hotplug related race conditions. I'm working on an alternative solution which would solve all of them.

**Comment by yarda** [comment]
Thanks, merged.

**Comment by olysonek** [comment]
I ended up using a part of this PR in my version, thanks!
https://github.com/redhat-performance/tuned/pull/167

**Pull request merged**